Dockerfile

```diff
@@ -0,0 +1,109 @@
+# ===============================================
+FROM helsinkitest/node:12-slim as appbase
+# ===============================================
+
+# Offical image has npm log verbosity as info. More info - https://github.com/nodejs/docker-node#verbosity
+ENV NPM_CONFIG_LOGLEVEL warn
+
+# Global npm deps in a non-root user directory
+ENV NPM_CONFIG_PREFIX=/app/.npm-global
+ENV PATH=$PATH:/app/.npm-global/bin
+
+# Set env
+ARG PORT
+ENV PORT $PORT
+
+# Use non-root user
+USER appuser
+
+# Copy package.json and package-lock.json/yarn.lock files
+COPY --chown=appuser:appuser package*.json *yarn* ./
+
+# Copy internals becaused they are called in preinstall and postinstall
+# hooks
+COPY --chown=appuser:appuser internals/ ./internals/
+
+# Install npm depepndencies
+ENV PATH /app/node_modules/.bin:$PATH
+
+# =============================
+FROM appbase as development_base
+# =============================
+
+# Set NODE_ENV to development in the development container
+ARG NODE_ENV=develop
+ENV NODE_ENV $NODE_ENV
+
+USER root
+RUN apt-install.sh \
+      build-essential \
+      autoconf \
+      libpng-dev
+
+USER appuser
+
+RUN npm install
+
+USER root
+RUN apt-cleanup.sh build-essential
+
+USER appuser
+
+# Copy all files
+COPY --chown=appuser:appuser . .
+
+# =============================
+FROM development_base as development
+# =============================
+
+# Bake package.json start command into the image
+CMD ["npm", "start"]
+
+# =============================
+FROM appbase as staticbuilder
+# =============================
+
+# Build scripts rely on devDependencies. When NODE_ENV is production,
+# some of these dependencies would not be installed.
+ARG NODE_ENV=develop
+ENV NODE_ENV $NODE_ENV
+
+USER root
+RUN apt-install.sh \
```

Comment:
If you want to create a smaller development image here you could chain the `apt-install.sh` and the `npm` command by doing something in the lines of

```
RUN apt-install.sh \
       build-essential \
       autoconf \
       libpng-dev && \
    su - appuser -c npm install && \
    apt-cleanup.sh build-essential
```

The problem with doing it this way is that the `apt-cleanup.sh` command will remove that files from the final image, but it will still have the cache from the apt-install command, so the final image will be just as large as if build-essential and the rest would still be installed.

Reply:
I'm installing dependencies twice in the Dockerfile. Should I use this same method for the production dependencies?

Actually, before you answer that, a bit of a sidetrack:
I currently install the same dependencies twice in this file. When in production, I'd prefer to only install `npm` deps that are required for the production version. But the current tooling relies on `devDependencies` for the production build, which means that I have to install all dependencies all the time.

I added two installs to make it easier for some dev at some point to be able to fix the situation, but I feel like that's not useful. Doesn't it currently increase build time by a factor?

Reply:
I think that this change fails the build. The error I am seeing at Gitlab is pretty generic: 
>docker.errors.BuildError: The command '/bin/sh -c apt-install.sh        build-essential        autoconf        libpng-dev &&     su - appuser -c npm install &&     apt-cleanup.sh build-essential' returned a non-zero code: 1
 Build failed!
 ERROR: Job failed: command terminated with exit code 1

Reply:
The command that I created on the fly, so it might contain errors. I would recommend trying to build locally and there see what is happening. The thing I'm doing in the script is basically just running one of the commands as another user.

`su - appuser -c "npm install"` could be tried for instance, with the command in quotes.

And regarding the double install of the dependencies. I see no difference to the two installation so might as well just use one. However, if you in development are using development NPM packages, than you should of course have two separate `npm install`s where one installs the main things to run the application while the other one (in the develpoment stage) only installs the dev packages.

Reply:
I think the culprit here ended up being the login flag (`-`). Because the flag was up, the command was ran in a new session within appuser's home (?) directory. This directory did not have the `package.json` and `package-lock.json` files which are needed for `npm install` to succeed.

I only updated the file so that it only installs dependencies once.